Wrong suggestion of `const fn` if type implements Drop

Hello together,

The following code snippet (in `src/lib.rs`):
```rust
pub struct A;
impl A {
    #![warn(clippy::missing_const_for_fn)]
    pub fn a(self) -> B {
        B
    }
}
impl Drop for A {
    fn drop(&mut self) {}
}

pub struct B;
```
leads to this suggestion:
```
warning: this could be a const_fn
 --> src/lib.rs:4:5
  |
4 | /     pub fn a(self) -> B {
5 | |         B
6 | |     }
  | |_____^
  |
note: lint level defined here
 --> src/lib.rs:3:13
  |
3 |     #![warn(clippy::missing_const_for_fn)]
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
```
but the suggested fix (i.e. add the `const` qualifier) is wrong. After the fix the following error occurs:
```
error[E0493]: destructors cannot be evaluated at compile-time
 --> src/lib.rs:4:20
  |
4 |     pub const fn a(self) -> B {
  |                    ^^^^ constant functions cannot evaluate destructors

error: aborting due to previous error
```

# Meta
```
$ cargo clippy -V
clippy 0.0.212 (e3cb40e 2019-06-25)

$ rustc --version
rustc 1.37.0 (eae3437df 2019-08-13)
```

--- a/clippy_lints/src/missing_const_for_fn.rs
+++ b/clippy_lints/src/missing_const_for_fn.rs
@@ -1,10 +1,11 @@
-use crate::utils::{is_entrypoint_fn, span_lint, trait_ref_of_method};
+use crate::utils::{has_drop, is_entrypoint_fn, span_lint, trait_ref_of_method};
 use rustc::hir;
 use rustc::hir::intravisit::FnKind;
-use rustc::hir::{Body, Constness, FnDecl, HirId};
+use rustc::hir::{Body, Constness, FnDecl, HirId, HirVec};
 use rustc::lint::{in_external_macro, LateContext, LateLintPass, LintArray, LintPass};
 use rustc::{declare_lint_pass, declare_tool_lint};
 use rustc_mir::transform::qualify_min_const_fn::is_min_const_fn;
+use rustc_typeck::hir_ty_to_ty;
 use syntax_pos::Span;
 
 declare_clippy_lint! {
@@ -94,7 +95,10 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingConstForFn {
                 }
             },
             FnKind::Method(_, sig, ..) => {
-                if trait_ref_of_method(cx, hir_id).is_some() || already_const(sig.header) {
+                if trait_ref_of_method(cx, hir_id).is_some()
+                    || already_const(sig.header)
+                    || method_accepts_dropable(cx, &sig.decl.inputs)
+                {
                     return;
                 }
             },
@@ -111,6 +115,16 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingConstForFn {
             span_lint(cx, MISSING_CONST_FOR_FN, span, "this could be a const_fn");
         }
     }
+}
+
+/// Returns true if any of the method parameters is a type that implements `Drop`. The method
+/// can't be made const then, because `drop` can't be const-evaluated.
+fn method_accepts_dropable(cx: &LateContext<'_, '_>, param_tys: &HirVec<hir::Ty>) -> bool {
+    // If any of the params are dropable, return true
+    param_tys.iter().any(|hir_ty| {
+        let ty_ty = hir_ty_to_ty(cx.tcx, hir_ty);
+        has_drop(cx, ty_ty)
+    })
 }
 
 // We don't have to lint on something that's already `const`

--- a/tests/ui/missing_const_for_fn/cant_be_const.rs
+++ b/tests/ui/missing_const_for_fn/cant_be_const.rs
@@ -68,3 +68,25 @@ impl std::ops::Add for Point {
         Point(self.0 + other.0, self.1 + other.1)
     }
 }
+
+mod with_drop {
+    pub struct A;
+    pub struct B;
+    impl Drop for A {
+        fn drop(&mut self) {}
+    }
+
+    impl A {
+        // This can not be const because the type implements `Drop`.
+        pub fn a(self) -> B {
+            B
+        }
+    }
+
+    impl B {
+        // This can not be const because `a` implements `Drop`.
+        pub fn a(self, a: A) -> B {
+            B
+        }
+    }
+}

--- a/tests/ui/missing_const_for_fn/could_be_const.rs
+++ b/tests/ui/missing_const_for_fn/could_be_const.rs
@@ -53,5 +53,20 @@ fn generic_arr<T: Copy>(t: [T; 1]) -> T {
     t[0]
 }
 
+mod with_drop {
+    pub struct A;
+    pub struct B;
+    impl Drop for A {
+        fn drop(&mut self) {}
+    }
+
+    impl B {
+        // This can be const, because `a` is passed by reference
+        pub fn b(self, a: &A) -> B {
+            B
+        }
+    }
+}
+
 // Should not be const
 fn main() {}

--- a/tests/ui/missing_const_for_fn/could_be_const.stderr
+++ b/tests/ui/missing_const_for_fn/could_be_const.stderr
@@ -49,5 +49,13 @@ LL | |     t
 LL | | }
    | |_^
 
-error: aborting due to 6 previous errors
+error: this could be a const_fn
+  --> $DIR/could_be_const.rs:65:9
+   |
+LL | /         pub fn b(self, a: &A) -> B {
+LL | |             B
+LL | |         }
+   | |_________^
+
+error: aborting due to 7 previous errors
 

